+sw_tests/+unit_tests/unittest_ndbase_optimisers.m

```diff
@@ -48,28 +57,27 @@ function test_optimise_rosen_upper_bound_minimum_accessible(testCase, optimiser)
         end
 
         function test_optimise_rosen_upper_bound_minimum_not_accessible(testCase, optimiser)
-            % note intital guess is outside bounds
             [pars_fit, cost_val, ~] = optimiser([], testCase.rosenbrock, [-1,-1], 'ub', [0, inf]);
             testCase.verify_val(pars_fit, [0, 0], 'abs_tol', 1e-3);
-            testCase.verify_val(cost_val, 1, 'abs_tol', 1e-6);
+            testCase.verify_val(cost_val, 1, 'abs_tol', 2e-3);
         end
 
         function test_optimise_rosen_both_bounds_minimum_accessible(testCase, optimiser)
-            [pars_fit, cost_val, ~] = optimiser([], testCase.rosenbrock, [-1,-1], 'lb', [-2, -2], 'ub', [2, 2]);
+            [pars_fit, cost_val, ~] = optimiser([], testCase.rosenbrock, [-1,-1], 'lb', [-5, -5], 'ub', [5, 5]);
             testCase.verify_val(pars_fit, testCase.rosenbrock_minimum, 'abs_tol', 1e-3);
             testCase.verify_val(cost_val, 0, 'abs_tol', 1e-6);
         end
 
-        function test_optimise_rosen_both_bounds_minimum_not_accessible(testCase, optimiser)
+        function test_optimise_rosen_both_bounds_minimum_not_accessible(testCase)
```

Comment:
Just checked and `lsqnonlin` does find the lowest cost val here at `(0,0)`
I think the problem could be solved by moving slightly away from the boundary edge when we reset invalid parameters outside bounds?

Reply:
Yeh if you change to have the intial guess `0.1` inside the `lb` like so
`'lb', [-1.1, -1.1]`
This test passes for `lm4` minimiser

Reply:
I just had a play and it still seems flaky. For some bounds it works and for others it doesn't... I wonder if we have to transform the `diff_step` too?

Reply:
I think the diff step has to be taken in the transformed coordinate so as not to go over the bounds -  the diff step is relative - i.e. scaled by the magnitude of the scaled/free parameter rather than the bound one which I also think it correct. We can discuss tomorrow!

Reply:
If you set `lsqnonlin` to have `MaxIteration=0` - then you can see the initial parameters are actually at the mid-point
```
[pars_lsq, cost_lsq] = lsqnonlin(rosenbrock,[-1, -1], [-0.5, -0.5], [0 0], optimoptions('lsqnonlin','MaxIterations', 0));
pars_lsq  %  [ 0.25, 0.25]
```
If you do one-sided bound on the last parameter it does something else...
```
[pars_lsq, cost_lsq] = lsqnonlin(rosenbrock,[-1, -1], [-0.5, -0.5], [0 inf], optimoptions('lsqnonlin','MaxIterations', 0));
pars_lsq  %  [    -0.2500, 0]
```
Perhaps it does this (rather than setting the parameter to the nearest bound) for a similar reason?

Reply:
So the Matlab code which does the start parameter modification is in `/toolbox/shared/optimlib/startx.m` and does a box-centered start - if the input `x0` is out of bounds and:

* when both bounds are specified it sets the input to be the mean of the upper and lower bound: `x0 = (ub + lb) / 2`
* when only the upper or lower bound is specified it sets it to the bound +/- 1 or half the bound whichever is greater: `x0 = ub - max(ub/2, 1)` or `x0 = lb + max(lb/2, 1)`.

Maybe we should do the same for `lm4`?

Reply:
Yeh agreed, thanks for looking into it - I think if they've chosen that as a default we'd be unlikely to find a better one (user can't complain too much if initial parameter out of bounds).

FYI this change would apply to all minimisers as it's done in the cost function wrapper.

Reply:
I think that would be fine - I guess you're going to change `get_free_parameters`?

Reply:
So with implementing the box start parameters in [this gist](https://gist.github.com/mducle/becf9f475489a9b2baadf661eb12a92c) I could get all the tests to pass with for both `simplex` and `lm4`, even with the original bounds in `both_bounds_minimum_accessible` and with a tighter tolerance for `upper_bound_minimum_not_accessible` (`1e-4` compared to `1e-6` for `simplex`), but I had to change the default `nu_up` from `10` to `5` in order to get those two tests to fit with the original bounds and tighter tolerance... not sure if this is ok?

Reply:
Thanks @mducle - I'll look at this today - in theory I don't have a problem with changing `nu_up=5` as long as it still passes the NIST data sets. 
Here is the advice given on wikipedia...

> An effective strategy for the control of the damping parameter, called delayed gratification, consists of increasing the parameter by a small amount for each uphill step, and decreasing by a large amount for each downhill step. The idea behind this strategy is to avoid moving downhill too fast in the beginning of optimization, therefore restricting the steps available in future iterations and therefore slowing down convergence.[[7]](https://en.wikipedia.org/wiki/Levenberg%E2%80%93Marquardt_algorithm#cite_note-Transtrum2011-7) An increase by a factor of 2 and a decrease by a factor of 3 has been shown to be effective in most cases, while for large problems more extreme values can work better, with an increase by a factor of 1.5 and a decrease by a factor of 5.[[8]](https://en.wikipedia.org/wiki/Levenberg%E2%80%93Marquardt_algorithm#cite_note-Transtrum2012-8)

Which I think (if I'm interpreting it correctly) would correspond to `nu_up=2, nu_dn=0.33` (similar value to existing default value for `nu_dn`). I will try `nu_up = 10 (existing default), 5, 2` and see how they compare in terms of the number of iterations as well as accuracy. FYI I got my value of `nu_up` from numerical recipes (or something based on numerical recipes...).

Reply:
So I quickly tried this on the `Gauss3` NIST dataset (so-called average difficulty) and get these results
```
% minimising residual array
% nu_dn   nu_up   iterations  nhess_evals     dcost(%)
% 10        0.3      11          10             -1.1e-9
% 5         0.3       9           8             -1.1e-9
% 2         0.3      11           8             -1.1e-9
% 1.5       0.2      15           8             -1.1e-9

% minimising scalar
% nu_dn   nu_up   iterations  nhess_evals     dcost(%)
% 10        0.3      12         10            -1.2e-7
% 5         0.3      17          8            -6.5e-8
% 2         0.3      16         10            -1.2e-7
% 1.5       0.2      21          9            -1.2e-7
```
![image](https://github.com/user-attachments/assets/9d2a8b21-121d-413e-bfb0-75221bcde897)
I will try on a more difficult dataset...

Reply:
Seems like `nu_up=5, nu_dn=0.3` is a bit of a sweet spot! Albeit having not given it the rigorous fitbenchmarking treatment...